interfaces/builtin/kubernetes_support.go

```diff
@@ -35,65 +47,227 @@ const kubernetesSupportBaseDeclarationSlots = `
+  /bin/true ixr,
+  ptrace trace peer=unconfined,
+###KUBERNETES_SUPPORT_SYSTEMD_RUN###
+}
+`
+
+const kubernetesSupportConnectedPlugAppArmorKubelet = `
+# Allow running as the kubelet service
+
+# Ideally this would be snap-specific
+/run/dockershim.sock rw,
+
+# allow managing pods' cgroups
+/sys/fs/cgroup/*/kubepods/{,**} rw,
 
-@{PROC}/sys/kernel/panic rw,
-@{PROC}/sys/kernel/panic_on_oops rw,
+# Allow tracing our own processes. Note, this allows seccomp sandbox escape on
+# kernels < 4.8
+capability sys_ptrace,
+ptrace (trace) peer=snap.@{SNAP_NAME}.*,
+
+# Allow ptracing other processes (as part of ps-style process lookups). Note,
+# the peer needs a corresponding tracedby rule. As a special case, disallow
+# ptracing unconfined.
+ptrace (trace),
+deny ptrace (trace) peer=unconfined,
+
+@{PROC}/[0-9]*/attr/ r,
+@{PROC}/[0-9]*/fdinfo/ r,
+@{PROC}/[0-9]*/map_files/ r,
+@{PROC}/[0-9]*/ns/ r,
+
+@{PROC}/sys/kernel/panic w,
+@{PROC}/sys/kernel/panic_on_oops w,
 @{PROC}/sys/kernel/keys/root_maxbytes r,
 @{PROC}/sys/kernel/keys/root_maxkeys r,
-@{PROC}/sys/vm/overcommit_memory rw,
-@{PROC}/sys/vm/panic_on_oom r,
 
-@{PROC}/diskstats r,
-@{PROC}/@{pid}/cmdline r,
+/dev/kmsg r,
+
+# kubelet calls out to systemd-run for some mounts, but not all of them and not
+# unmounts...
+capability sys_admin,
+mount /var/snap/@{SNAP_NAME}/common/{,**} -> /var/snap/@{SNAP_NAME}/common/{,**},
+mount options=(rw, rshared) -> /var/snap/@{SNAP_NAME}/common/{,**},
+
+/bin/umount ixr,
+deny /run/mount/utab rw,
+umount /var/snap/@{SNAP_INSTANCE_NAME}/common/**,
+`
+
+const kubernetesSupportConnectedPlugAppArmorKubeletSystemdRun = `
+  # kubelet mount rules
+  capability sys_admin,
+  /bin/mount ixr,
+  mount -> /var/snap/@{SNAP_INSTANCE_NAME}/common/**,
+  deny /run/mount/utab rw,
+`
+
+const kubernetesSupportConnectedPlugSeccompKubelet = `
+# Allow running as the kubelet service
+mount
+umount
+umount2
+
+unshare
+setns - CLONE_NEWNET
+`
+
+var kubernetesSupportConnectedPlugUDevKubelet = []string{
+	`KERNEL=="kmsg"`,
+}
+
+const kubernetesSupportConnectedPlugAppArmorKubeproxy = `
+# Allow running as the kubeproxy service
+
+# managing our own cgroup
+/sys/fs/cgroup/*/kube-proxy/{,**} rw,
 
 # Allow reading the state of modules kubernetes needs
+/sys/module/libcrc32c/initstate r,
 /sys/module/llc/initstate r,
 /sys/module/stp/initstate r,
+/sys/module/ip_vs/initstate r,
+/sys/module/ip_vs_rr/initstate r,
+/sys/module/ip_vs_sh/initstate r,
+/sys/module/ip_vs_wrr/initstate r,
+`
 
-# Allow listing kernel modules. Note, seccomp blocks module loading syscalls
-/sys/module/apparmor/parameters/enabled r,
-/bin/kmod ixr,
-/etc/modprobe.d/{,**} r,
+var kubernetesSupportConnectedPlugKmodKubeProxy = []string{
+	`ip_vs_rr`,
+	`ip_vs_sh`,
+	`ip_vs_wrr`,
+	`libcrc32c`,
+	`llc`,
+	`stp`,
+}
 
-# Allow ptracing Docker containers
-ptrace (read, trace) peer=docker-default,
-ptrace (read, trace) peer=snap.docker.dockerd,
+type kubernetesSupportInterface struct{}
```

Comment:
Note that embedding commonInterface would let you avoid some of the boilerplate. Did you consider doing that but decided not to or would you be open to simplifying this?

Reply:
I converted from commonInterface so I could more readily do the attribute handling. If there is an easier way, sure, I'm up for it being simplified.

Reply:
Done